prediction_market_agent/tools/betting_strategies.py

```diff
@@ -0,0 +1,215 @@
+from prediction_market_agent.data_models.market_data_models import Market
+from prediction_market_agent.tools.gnosis_rpc import GNOSIS_RPC_URL
+from prediction_market_agent.tools.web3_utils import xdai_to_wei, wei_to_xdai
+from prediction_market_agent.tools.types import xDai
+
+OutcomeIndex = t.Literal[0, 1]
+
+
+def get_market_moving_bet(
+    market_address: str,
+    target_p_yes: float,
+    max_iters: int = 100,
+    check_vs_contract: bool = False,  # Disable by default, as it's slow
+    verbose: bool = False,
+) -> t.Tuple[xDai, OutcomeIndex]:
+    """
+    Implements a binary search to determine the bet that will move the market's
+    `p_yes` to that of the target.
+
+    Consider a binary fixed-product market containing `x` and `y` tokens.
+    A trader wishes to aquire `x` tokens by betting an amount `d0`.
+
+    The calculation to determine the number of `x` tokens he acquires, denoted
+    by `dx`, is:
+
+    a_x * a_y = fixed_product
+    na_x = a_x + d0
+    na_y = a_y + d0
+    na_x * na_y = new_product
+    (na_x - dx) * na_y = fixed_product
+    (na_x * na_y) - (dx * na_y) = fixed_product
+    new_product - fixed_product = dx * na_y
+    dx = (new_product - fixed_product) / na_y
+    """
+    web3 = Web3(Web3.HTTPProvider(GNOSIS_RPC_URL))
+    market: Market = get_market(market_address)
+    amounts = market.outcomeTokenAmounts
+    prices = market.outcomeTokenMarginalPrices
+    if len(amounts) != 2 or len(prices) != 2:
+        raise ValueError("Only binary markets are supported.")
+
+    fixed_product = reduce(lambda x, y: x * y, amounts, 1)
+    assert np.isclose(float(sum(prices)), 1)
+    current_p_yes = prices[0]
+    bet_outcome_index = 0 if target_p_yes > current_p_yes else 1
+
+    min_bet_amount = 0
+    max_bet_amount = 100 * sum(amounts)  # TODO set a better upper bound
+
+    # Binary search for the optimal bet amount
+    for _ in range(max_iters):
+        bet_amount = (min_bet_amount + max_bet_amount) / 2
+        bet_amount_ = bet_amount * (xdai_to_wei(1) - market.fee) / xdai_to_wei(1)
+
+        # Initial new amounts are old amounts + equal new amounts for each outcome
+        amounts_diff = bet_amount_
+        new_amounts = [amounts[i] + amounts_diff for i in range(len(amounts))]
+
+        # Now give away tokens at `bet_outcome_index` to restore invariant
+        new_product = reduce(lambda x, y: x * y, new_amounts, 1)
+        dx = (new_product - fixed_product) / new_amounts[1 - bet_outcome_index]
+
+        # Sanity check the number of tokens against the contract
+        if check_vs_contract:
+            expected_trade = omen_calculate_buy_amount(
+                web3=web3,
+                market=market,
+                investment_amount=int(bet_amount),
+                outcome_index=bet_outcome_index,
+            )
+            assert np.isclose(
+                float(expected_trade) / dx, 1, rtol=0.01
+            )  # TODO always 1% off, investigate
```

Comment:
Maybe because of

```
# Allow 1% slippage.
min_outcome_tokens_to_buy = remove_fraction(calculated_shares, 0.01)
```

inside of `omen_calculate_buy_amount`? :( Sorry! 

The reason for that was that `omen_calculate_buy_amount` is used in buying the shares, and it could happen that the market moved a little in the meantime between calling this function and the buy function.

I'll modify the  `omen_calculate_buy_amount` to not have the default 1% slippage.

Reply:
Fixing here: https://github.com/evangriffiths/prediction-market-agent/pull/12

Reply:
Yay, that fixed it!